File: web/src/lib/lectura.ts
import type { Health } from './api';

/**
 * Con qué se transcribe el papel de este trabajo.
 *
 * No es lo mismo que elegir modelo en `oracles.ts`, y la diferencia importa al
 * elegir. Allí se decide quién juzga una costura que la estructura no pudo
 * resolver; acá, con qué se lee la imagen, que es el paso anterior a todo lo
 * demás. Un expediente mal leído no lo arregla ningún juez de costuras porque no
 * queda nada sobre lo que juzgar.
 *
 * Es opcional porque cuesta dinero y no siempre hace falta. Lo impreso lo lee
 * gratis el motor local; lo escrito a mano no lo lee, y en los libros de
 * registro antiguos todo lo que identifica al documento -- el folio, el nombre,
 * la cédula, el título, la fecha -- está escrito a mano sobre un formulario
 * impreso. Quién tiene delante cuál de los dos casos lo sabe el operador.
 */
export type LecturaChoice = 'local' | 'mistral';

export interface LecturaOption {
  id: LecturaChoice;
  label: string;
  hint: string;
  /** Si el servicio declara tener llave para pedirlo. */
  available: boolean;
  /** Por qué no se puede pedir, o `null` cuando sí. */
  reason: string | null;
}

const CATALOG: { id: LecturaChoice; label: string; hint: string }[] = [
  {
    id: 'local',
    label: 'Tesseract',
    hint: 'lo de siempre: gratis y suficiente para lo impreso'
  },
  {
    id: 'mistral',
    label: '+ Mistral OCR',
    hint: 'lee lo escrito a mano; se paga sólo lo que Tesseract no supo leer'
  }
];

const SIN_LLAVE = 'No hay llave de Mistral configurada en el servicio.';

const SIN_SERVICIO =
  'El servicio no informó con qué puede leer. Puede pedirse el motor local.';

/**
 * Los motores que la pantalla puede ofrecer, con el motivo de los que no.
 *
 * Mismo criterio que con los modelos: nunca se adivina disponibilidad. Ofrecer
 * un motor que no se puede pedir gasta la decisión del operador y la contesta
 * con un 422 cuando ya había elegido el archivo.
 */
export function lecturaOptions(health: Health | null): LecturaOption[] {
  const declared = health?.readers ?? null;
  return CATALOG.map((entry) => {
    if (entry.id === 'local') {
      return { ...entry, available: true, reason: null };
    }
    if (!declared) {
      return { ...entry, available: false, reason: SIN_SERVICIO };
    }
    const available = declared[entry.id] === true;
    return { ...entry, available, reason: available ? null : SIN_LLAVE };
  });
}

/** Por qué no se puede pedir lo que está elegido, o `null` si sí se puede. */
export function reasonForLectura(
  choice: LecturaChoice,
  options: LecturaOption[]
): string | null {
  const option = options.find((candidate) => candidate.id === choice);
  if (!option) return SIN_SERVICIO;
  return option.available ? null : option.reason;
}

File: backend/src/resolutions/application/lectura.py
"""Con qué se lee el papel, que es una decisión del operador y no del servicio.

Distinta de la de `oracle.py`, y conviene no confundirlas porque contestan
preguntas distintas. Allí se elige **quién juzga una costura dudosa**: un modelo
de razonamiento mirando huellas de páginas que ya se leyeron. Aquí se elige
**con qué se transcribe la imagen**, que es el paso anterior a todo lo demás. Un
expediente mal leído no se arregla con un buen juez de costuras: no hay nada
sobre lo que juzgar.

Por qué es opcional y no automático: el motor local es gratis y el remoto se
paga por página. En un expediente mecanografiado el remoto no aporta nada --
Tesseract lee lo impreso perfectamente -- y activarlo sería pagar por leer dos
veces lo mismo. En un libro de registro de 1982 aporta lo único que importa,
porque ahí todo lo que identifica al documento está escrito a mano: el folio, el
nombre del graduado, la cédula, el título y la fecha. Quién sabe cuál de los dos
casos tiene delante es el operador, y por eso la decisión es suya y viaja con el
trabajo, igual que la acción y el modelo de bordes.

La misma regla que sostiene la elección de modelo vale acá: **una elección no se
degrada en silencio**. Pedir el motor de pago sin su llave es un error que se
contesta antes de aceptar la subida, no una sustitución callada que alguien
descubre tres cajas más tarde, cuando ya archivó lo que salió.
"""

from __future__ import annotations

from collections.abc import Mapping
from enum import StrEnum


class LecturaChoice(StrEnum):
    """Con qué motor se transcribe cada imagen de este trabajo."""

    #: Tesseract y nada más. Es el valor por omisión, porque es el que había y
    #: porque es el que no cuesta dinero: quien no pida otra cosa sigue
    #: trabajando exactamente como trabajaba.
    LOCAL = "local"

    #: Tesseract primero y, donde el local no saque nada legible, el OCR de
    #: documentos de Mistral. En ese orden y no al revés: lo impreso lo lee
    #: gratis el de casa, y lo que se paga es sólo lo que el de casa no supo
    #: leer, que en estos archivos es lo escrito a mano.
    MISTRAL = "mistral"

    @property
    def label(self) -> str:
        return _LABELS[self]

    @property
    def key_setting(self) -> str | None:
        """Qué ajuste del worker tiene que traer llave. `None` para el local."""
        return _KEY_SETTINGS.get(self)

    @property
    def env_var(self) -> str | None:
        """Qué variable de entorno nombrar cuando la llave falta."""
        return _ENV_VARS.get(self)

    def is_available(self, settings: Mapping[str, object]) -> bool:
        if self is LecturaChoice.LOCAL:
            return True
        setting = self.key_setting
        return bool(setting and settings.get(setting))

    @classmethod
    def parse(cls, value: str | None) -> LecturaChoice:
        """Interpreta lo que llegó por la API. Lo desconocido no se adivina."""
        if value in (None, ""):
            return cls.LOCAL
        try:
            return cls(str(value).strip().lower())
        except ValueError as error:
            opciones = ", ".join(choice.value for choice in cls)
            raise ValueError(
                f"motor de lectura desconocido: {value!r}; se esperaba uno de {opciones}"
            ) from error


def available_readers(settings: Mapping[str, object]) -> dict[str, bool]:
    """De cada motor, si hay llave para pedirlo.

    Va en `/api/health` por el mismo motivo que la de los modelos: para que la
    pantalla apague lo que no se puede pedir y diga por qué, en vez de ofrecerlo
    y cosechar un 422 cuando el operador ya eligió el archivo.
    """
    return {choice.value: choice.is_available(settings) for choice in LecturaChoice}


_LABELS = {
    LecturaChoice.LOCAL: "Tesseract",
    LecturaChoice.MISTRAL: "Tesseract + Mistral OCR",
}

_KEY_SETTINGS = {LecturaChoice.MISTRAL: "mistral_api_key"}

_ENV_VARS = {LecturaChoice.MISTRAL: "MISTRAL_API_KEY"}
